man/tapelist.5
'\" t
.\"     Title: tapelist
.\"    Author: James da Silva <jds@amanda.org>
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 07/25/2012
.\"    Manual: File formats and conventions
.\"    Source: Amanda 3.3.2
.\"  Language: English
.\"
.TH "TAPELIST" "5" "07/25/2012" "Amanda 3\&.3\&.2" "File formats and conventions"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
tapelist \- The list of Amanda volumes in use
.SH "DESCRIPTION"
.PP
The
\fItapelist\fR
file contains the list of tapes in active use\&. This file is not ordinarily ordinarily edited or examined manually\&. Its format may change, or it may be removed altogether, in future versions of Amanda\&. It contains lines of the form:
.PP
.nf
YYYYMMDD label flags [BARCODE:barcode] [META:meta] [#comment]
.fi
.PP
Where
\fIYYYYMMDD\fR
is the date the tape was written,
\fIlabel\fR
is a label for the tape as written by
\fBamlabel\fR
and
\fIflags\fR
is one of "reuse" or "no\-reuse" and tells Amanda whether the tape may be reused\&. A non\-reusable tape will not be overwritten, even if it is older than
\fBdumpcycle\fR
dumps\&. See the
\fBreuse\fR
options of
\fBamadmin\fR(8)\&. The optional barcode is prefixed with \*(AqBARCODE\*(Aq if the volume have a barcode\&. The optional meta label is prefixed with \*(AqMETA\*(Aq if the volume is on a meta volume\&. The optional comment is prefixed with a \*(Aq#\*(Aq and continues to the end of the line\&. Amanda will maintain this comment until the tape is overwritten\&. Note that comments are not allowed on blank lines\&.
.PP
The file is sorted such that the most\-recently used tape appears on the first line, and the oldest tape appears on the last line\&.
.PP
\fBAmdump\fR
and
\fBamflush\fR
will by default refuse to write to an unlabeled tape, or to a labeled tape that is considered active or marked "no\-reuse"\&. There must be more tapes in active rotation (see the
\fBtapecycle\fR
option) than there are runs in the backup cycle (see the
\fBdumpcycle\fR
option) to prevent overwriting a backup image that would be needed to do a full recovery\&.
.SH "SEE ALSO"
.PP
\fBamanda\fR(8),
\fBamanda.conf\fR(5)
.PP
The Amanda Wiki:
: http://wiki.zmanda.com/
.SH "AUTHORS"
.PP
\fBJames da Silva\fR <\&jds@amanda\&.org\&>
.PP
\fBStefan G\&. Weichinger\fR <\&sgw@amanda\&.org\&>
.PP
\fBDustin J\&. Mitchell\fR <\&dustin@zmanda\&.com\&>
.RS 4
Zmanda, Inc\&. (http://www\&.zmanda\&.com)
.RE


NEWS
Changes in release 2.5.0p2

* Fix major bug in amandad.
* amverifyeun works.
* amstatus works with ':' in diskname.

Changes in release 2.5.0p1

* Add the 'amtape update' command.
* Many small bug fix.

Changes in release 2.5.0

* Communication security/authentication: Kerberos 4/5, OpenSSH
* Data security: Symmetric/Assymetric encrytion algorithms (aesutil
  and gpg encryption),  Encryption can be done on server or client,
  Custom encryption utilities can be used.
* Compression: Ability to add custom compression tools. This is a
  really useful feature especially since it can specified for a DLE.  You
  can use different compression algorithm for images, binaries, ascii files
  and so on.
* Dump images spanning multiple media volumes - Dump images are no
  longer restricted to a single media volume (tape or vtape).  Data restoration
  can be done using amrecover and amfetchdump commands.
* Auto tape labelling - This optional feature is good for disk backups.

Change in release 2.4.5p1

* New [--days <num>] argument to 'amadmin <conf> tape'.
* New amoverview -last -num0 -togo0 options.

Change in release 2.4.5

* new displayunit global option to select the unit use to display number
  k=kilo, m=mega, g=giga, t=tera.
* new amoverview -skipmissed option.

Change in release 2.4.5b1

* holding disk disk use timestamped directory.
* autoflush flush today's dump.
* new bumppercent global option, this should improve bumping criteria,
  the bumpsize is set to a fixed value which can be adequate for small
  and large disk.
* bumpsize, bumppercent, bumpdays and bumpmult can be in a dumptype.
* calcsize support include and exclude like gnutar.
* new 'estimate' dumptype option to select estimate type:
    CLIENT: estimate by the dumping program.
    CALCSIZE: estimate by the calcsize program, a lot faster but less acurate.
    SERVER: estimate based on statistic from previous run, take second but
            can be wrong on the estimate size.

Changes in release 2.4.4p3

* new chg-iomega changer script.
* amanda will not use a tape if it's label is not in the tapelist file
  (eg. after an amrmtape).

Changes in release 2.4.4p2

* new initial_poll_delay config in chg-zd-mtx
* 'amadmin find' list disk removed from the disklist.
* amrecover can recover a disk removed from the disklist file.
* amanda works with samba3.
* new chg-disk changer script to use with the file: driver.
* Kerberos 4 support integrated and functional again.

Changes in release 2.4.4p1

* amverify do not advance to next tape at the end.
* new amflush -b and -s options, change -f behavior.
* new chg-mcutil changer.
* amrecover_changer works with chg-multi.
* default driveslot for chg-zd-mtx is now 0.
* amplot generate color postscript by default.
* amplot -b flag to generate b/w postscript.
* rait: works with file:
* print a postscript label for each tape used.

Changes in release 2.4.4

* New -a option to amcheck to always send an email.

Changes in release 2.4.4b1

* maxpromoteday: New option for a dumptype.
* New amtapetype program (replace tapetype)
* Client compile on CYGWIN
* amrestore -f <fileno> : amrestore can do fsf before reading the tape
* amrestore -l <label>  : amrestore can check the label before restoring.
* New config options:
  amrecover_do_fsf: Amrecover will use the -f flag of amrestore to position the
                    tape.
  amrecover_check_label: amrecover will pass the -l flag to amrestore.
  amrecover_changer: amrecover will use your changer.
  maxdumpsize: The maximum size of the dumps during a run,
               default to tapesize * runtapes
  taperalgo: select your algo for the taper, try 'largestfit', it can improve
             tape usage.
* amrecover allow to retry a skip a tape.
* New --days option to 'amadmin <config> balance' command.
* New --date option to amstatus
* amreport print an usage by tape.
* Newer promote algorithm that try to reduce the number of full on a
  single host in a run.
* New changer: chg-juke, chg-null and chg-rait

Changes in release 2.4.3

* Few small bug fixed
* New runtapes argument to amverify.
* New amverifyrun program.
* 2.4.3 client works with older server.

Changes in release 2.4.3b4

* new noop packet type. It allow the server to know the client features.
* exclude list with relative path is working again.
* amstatus is working again.

Changes in release 2.4.3b3

* --with-maxtapeblocksize configure options
* blocksize tapetype option
* file-pad tapetype option
* Multiple exclude in dumptype
* Option include in dumptype
* New disklist syntax:
*   hostname diskname [ diskdevice ] dumptype [ spindle [ interface ] ]
* chg-zd-mtx: Major cleanup and general overhaul.
* amrecover: new listdisk command.

Changes in release 2.4.3b2

* amrestore try to find the next chunk in the current directory.
* amrecover: You can change the tape device between each tape.
* amstatus work with amflush.

Changes in release 2.4.3b1

* tapeio.
* New [host [disk]*]* arguments to amdump, amcheck, amadmin and amflush.
* New [-D datastamp]* arguments to amflush.
* amrecover: cd accept shell wildcard, new command cdx accept regex.
* new autoflush option.
* new dumperstr option to specify the priority order of each dumper.

Changes in release 2.4.2

* Samba passwords are now sent to smbclient via a pipe and never displayed.
* Debug files in /tmp/amanda (--with-debug-dir) are now timestamped and old
  ones automatically cleaned out.  This means more space (a few KBytes) will
  be used since in a given run, several of the programs are called more than
  once.  But it also means important debugging information should no longer
  be lost by the file being overwritten.  The length of time to keep the
  files is controlled by --with-debug-days (default: 4).  The old flag
  --with-pid-debug-files is no longer needed and is ignored.
* dumptypes can be specialized in the disklist file
* Amanda will never use more space than specified (use) for an holding
  disk

perl/Amanda/Tapelist.pm
# This file was automatically generated by SWIG (http://www.swig.org).
# Version 2.0.4
#
# Do not make changes to this file unless you know what you are doing--modify
# the SWIG interface file instead.

package Amanda::Tapelist;
use base qw(Exporter);
use base qw(DynaLoader);
package Amanda::Tapelistc;
bootstrap Amanda::Tapelist;
package Amanda::Tapelist;
@EXPORT = qw();

# ---------- BASE METHODS -------------

package Amanda::Tapelist;

sub TIEHASH {
    my ($classname,$obj) = @_;
    return bless $obj, $classname;
}

sub CLEAR { }

sub FIRSTKEY { }

sub NEXTKEY { }

sub FETCH {
    my ($self,$field) = @_;
    my $member_func = "swig_${field}_get";
    $self->$member_func();
}

sub STORE {
    my ($self,$field,$newval) = @_;
    my $member_func = "swig_${field}_set";
    $self->$member_func($newval);
}

sub this {
    my $ptr = shift;
    return tied(%$ptr);
}


# ------- FUNCTION WRAPPERS --------

package Amanda::Tapelist;

*get_last_reusable_tape_label = *Amanda::Tapelistc::get_last_reusable_tape_label;
*list_new_tapes = *Amanda::Tapelistc::list_new_tapes;
*C_read_tapelist = *Amanda::Tapelistc::C_read_tapelist;
*C_clear_tapelist = *Amanda::Tapelistc::C_clear_tapelist;

# ------- VARIABLE STUBS --------

package Amanda::Tapelist;


@EXPORT_OK = ();
%EXPORT_TAGS = ();


=head1 NAME

Amanda::Tapelist - manipulate the Amanda tapelist

=head1 SYNOPSIS

    use Amanda::Tapelist;

    # to get a read only copy of the tapelist file:
    my $tl = Amanda::Tapelist->new("/path/to/tapefile");

    # to read/update/write the tapelist file
    # read and take lock
    my $tl = Amanda::Tapelist->new("/path/to/tapefile", 1);
    # modify the memory copy
    $tl->add_tapelabel($datestamp, $label);
    $tl->add_tapelabel($datestamp2, $label2, $comment, 1);
    # write it and unlock
    $tl->write();

    # If you already have a read only copy and want to modify it
    # take a read only copy
    my $tl = Amanda::Tapelist->new("/path/to/tapefile");
    # reload and take lock
    $tl->reload(1);
    # modify the memory copy
    tl->add_tapelabel($datestamp, $label);
    $tl->add_tapelabel($datestamp2, $label2, $comment, 1);
    # write it and unlock
    $tl->write();

=head1 OBJECT-ORIENTED INTERFACE

C<new> returns a hash with no C<tles> set if the tapelist does
not exist. C<tles> is an empty array if the tapelist is empty.
Invalid entries are silently ignored.

=head2 tapelist object

A tapelist object is a hash with the following keys:

=over

=item C<filename>

  The filename of the tapelist file.

=item C<filename_lock>

  The filename of the lock file.

=item C<fl>

  A Amanda::Util::file_lock is the file is locked.

=item C<tles>

A sequence of tapelist elements (referred to as TLEs in this document),
sorted by datestamp from newest to oldest.

=back

=head2 tapelist element

A tapelist elementas a hash with the following keys:

=over

=item C<position>

the one-based position of the TLE in the tapelist

=item C<datestamp>

the datestamp on which this was written, or "0" for an unused tape

=item C<reuse>

true if this tape can be reused when it is no longer active

=item C<label>

tape label

=item C<comment>

the comment for this tape, or undef if no comment was given

=back

=head1 Method

The following methods are available on a tapelist object C<$tl>:

=over

=item C<relod($lock)>

reload the tapelist file, lock it if $lock is set

=item C<lookup_tapelabel($lbl)>

look up and return a reference to the TLE with the given label

=item C<lookup_tapepos($pos)>

look up and return a reference to the TLE in the given position

=item C<lookup_tapedate($date)>

look up and return a reference to the TLE with the given datestamp

=item C<remove_tapelabel($lbl)>

remove the tape with the given label

=item C<add_tapelabel($date, $lbl, $comment, $reuse)>

add a tape with the given date, label, comment and reuse to the end of the
tapelist. reuse can be 1 or undef for a reusable volume, it must be 0 for
a no-reusable volume.

=item C<write()> or C<write($filename)>

write the tapelist out to the same file as when read or to C<$filename> if it
is set, remove the lock if a lock was taken

=item C<unlock()>

remove the lock if a lock was taken

=item C<clear_tapelist()>

remove all tle from the tles.

=back

=head1 INTERACTION WITH C CODE

The C portions of Amanda treat the tapelist as a global variable,
while this package treats it as an object (and can thus handle more
than one tapelist simultaneously).  Every call to C<reload>
fills this global variable with a copy of the tapelist, and likewise
C<clear_tapelist> clears the global.  However, any changes made from
Perl are not reflected in the C copy, nor are changes made by C
modules reflected in the Perl copy.

=cut



use Amanda::Debug qw(:logging);
use Amanda::Config qw( config_dir_relative );
use File::Copy;
use Fcntl qw(:flock); # import LOCK_* constants

## package functions

sub new {
    my ($class)  = shift;
    my ($filename, $lock ) = @_;
    my $self = {
	filename => $filename,
	lockname => $filename . '.lock',
    };
    bless $self, $class;

    $self->reload($lock);
    return $self;
}

sub clear_tapelist {
    my $self = shift;

    # clear the C version
    C_clear_tapelist();

    $self->{'tles'} = [];

    return $self;
}

## methods

sub reload {
    my $self = shift;
    my ($lock) = @_;

    if ($lock) {
	$self->_take_lock();
    }

    # clear the C copy
    C_clear_tapelist();

    # let C read the file
    C_read_tapelist($self->{'filename'});

    $self->_read_tapelist();
}

sub lookup_tapelabel {
    my $self = shift;
    my ($label) = @_;

    for my $tle (@{$self->{'tles'}}) {
	return $tle if ($tle->{'label'} eq $label);
    }

    return undef;
}

sub lookup_tapepos {
    my $self = shift;
    my ($position) = @_;

    $self->_update_positions();
    return $self->{'tles'}->[$position-1];
}

sub lookup_tapedate {
    my $self = shift;
    my ($datestamp) = @_;

    for my $tle (@{$self->{'tles'}}) {
	return $tle if ($tle->{'datestamp'} eq $datestamp);
    }

    return undef;
}

sub remove_tapelabel {
    my $self = shift;
    my ($label) = @_;

    for (my $i = 0; $i < @{$self->{tles}}; $i++) {
	if ($self->{tles}->[$i]->{'label'} eq $label) {
	    splice @{$self->{tles}}, $i, 1;
	    $self->_update_positions();
	    return;
	}
    }
}

sub add_tapelabel {
    my $self = shift;
    my ($datestamp, $label, $comment, $reuse, $meta, $barcode, $blocksize) = @_;
    $reuse = 1 if !defined $reuse;

    # prepend this (presumably new) volume to the beginning of the list
    my $tle = {
        'datestamp' => $datestamp,
        'label'     => $label,
        'reuse'     => $reuse,
        'barcode'   => $barcode,
        'meta'      => $meta,
        'blocksize' => $blocksize,
        'comment'   => $comment,
    };
    my $tles = $self->{'tles'};
    if (!defined $tles->[0] ||
	$tles->[0]->{'datestamp'} le $datestamp) {
	unshift @{$tles}, $tle;
    } elsif (defined $tles->[0] &&
	$tles->[@$tles-1]->{'datestamp'} gt $datestamp) {
	push @{$tles}, $tle;
    } else {
	my $added = 0;
	for my $i (0..(@$tles-1)) {
	    if ($tles->[$i]->{'datestamp'} le $datestamp) {
		splice @{$tles}, $i, 0, $tle;
		$added = 1;
		last;
	    }
	}
	push @{$tles}, $tle if !$added;
    }
    $self->_update_positions();
}

sub write {
    my $self = shift;
    my ($filename) = @_;
    my $result = TRUE;
    $filename = $self->{'filename'} if !defined $filename;

    my $new_tapelist_file = $filename . "-new-" . time();

    open(my $fhn, ">", $new_tapelist_file) or die("Could not open '$new_tapelist_file' for writing: $!");
    for my $tle (@{$self->{tles}}) {
	my $datestamp = $tle->{'datestamp'};
	my $label = $tle->{'label'};
	my $reuse = $tle->{'reuse'} ? 'reuse' : 'no-reuse';
	my $barcode = (defined $tle->{'barcode'})? (" BARCODE:" . $tle->{'barcode'}) : '';
	my $meta = (defined $tle->{'meta'})? (" META:" . $tle->{'meta'}) : '';
	my $blocksize = (defined $tle->{'blocksize'})? (" BLOCKSIZE:" . $tle->{'blocksize'}) : '';
	my $comment = (defined $tle->{'comment'})? (" #" . $tle->{'comment'}) : '';
	$result &&= print $fhn "$datestamp $label $reuse$barcode$meta$blocksize$comment\n";
    }
    my $result_close = close($fhn);
    $result &&= $result_close;

    return if (!$result);

    unless (move($new_tapelist_file, $filename)) {
	die ("failed to rename '$new_tapelist_file' to '$filename': $!");
    }

    # re-read from the C side to synchronize
    C_read_tapelist($filename);

    $self->unlock();

    return undef;
}

sub unlock {
    my $self = shift;

    return if !exists $self->{'fl'};

    $self->{'fl'}->unlock();
    delete $self->{'fl'}
}

## private methods

sub _take_lock {
    my $self = shift;

    if (!-e $self->{'lockname'}) {
	open(my $fhl, ">>", $self->{'lockname'});
	close($fhl);
    }
    my $fl = Amanda::Util::file_lock->new($self->{'lockname'});
    while(($r = $fl->lock()) == 1) {
	sleep(1);
    }
    if ($r == 0) {
	$self->{'fl'} = $fl;
    }
}

sub _read_tapelist {
    my $self = shift;

    my @tles;
    open(my $fh, "<", $self->{'filename'}) or return $self;
    while (my $line = <$fh>) {
	my ($datestamp, $label, $reuse, $barcode, $meta, $blocksize, $comment)
	    = $line =~ m/^([0-9]*)\s([^\s]*)\s(reuse|no-reuse)\s*(?:BARCODE:([^\s]*))?\s*(?:META:([^\s]*))?\s*(?:BLOCKSIZE:([^\s]*))?\s*(?:\#(.*))?$/mx;
	next if !defined $datestamp; # silently filter out bogus lines
	push @tles, {
	    'datestamp' => $datestamp,
	    'label' => $label,
	    'reuse' => ($reuse eq 'reuse'),
	    'barcode' => $barcode,
	    'meta' => $meta,
	    'blocksize' => $blocksize,
	    'comment' => $comment,
	};
    }
    close($fh);

    # sort in descending order by datestamp, sorting on position, too, to ensure
    # that entries with the same datestamp stay in the right order
    $self->{'tles'} = \@tles;
    $self->_update_positions();
    @tles = sort {
	   $b->{'datestamp'} cmp $a->{'datestamp'}
	|| $a->{'position'} <=> $b->{'position'}
	} @tles;

    $self->{'tles'} = \@tles;

    # and re-calculate the positions
    $self->_update_positions(\@tles);
}

# update the 'position' key for each TLE
sub _update_positions {
    my $self = shift;
    my $tles = $self->{'tles'};
    for (my $i = 0; $i < scalar @$tles; $i++) {
	$tles->[$i]->{'position'} = $i+1;
    }
}

1;
